rfd/0004-workspace-and-prebuild-storage.md

```diff
@@ -0,0 +1,97 @@
+---
+authors: Hugo Dutka (contact@hugodutka.com)
+state: draft
+---
+
+# RFD 4 - Workspace and prebuild storage management
+
+## What
+
+Specify high-level workspace and prebuild storage management in Hocus using [the local block registry](./0001-block-registry.md) and a remote OCI registry.
+
+## Why
+
+Goals:
+
+1. Store data without duplication between workspaces that use the same prebuilds.
+2. Allow sharing prebuilds between users on different machines to avoid redundant builds.
+3. Enable fast workspace start-up, even without local caching.
+
+Storage backed by OCI images and powered by [overlaybd](https://github.com/containerd/overlaybd) with lazy pulling fits these requirements.
+
+## Details
+
+### Why OCI images?
+
+Using OCI images lets us benefit from the OCI ecosystem. In particular, we make heavy use of overlaybd, which underpins our local block registry. Overlaybd supports lazy pulling, which allows us to run a workspace without fully downloading its backing files, and then download the rest in the background. We can also use standardized OCI registries for long-term storage and transferring images between machines, eliminating the need for a custom solution.
+
+### Types of images Hocus needs
+
+Hocus needs to keep track of the following types of images:
+
+- git repository images
+- prebuild images, two types:
+  - code images (based on git repository images)
+  - os images
+- workspace images, two types:
+  - code images (based on prebuild code images)
+  - os images (based on prebuild os images)
+
+In total, there are 5 types of images Hocus manages.
```

Comment:
Why is the distinction between code images and os images required in prebuilds?

Reply:
To enable better caching of the git repository. The code image is based on a git repository image that's shared between all projects based on a given repository.